Using SSLCertificateSocketFactory for TLS handshake doesn't work externally for Android older than 5.0

Currently, internal Android projects are using SSLCertificateSocketFactory for TLS handshake, but since ALPN only works well with Android 5.0 or newer, they use NPN for Android 4.4 and older.

It works just because internal OKHttp has been hacked to support NPN. 

With the external OKHttp, it will failed as:
```java
com.google.common.util.concurrent.UncheckedExecutionException: io.grpc.StatusRuntimeException: INTERNAL: Failed starting transport
	at io.grpc.stub.Calls.getUnchecked(Calls.java:117)
	at io.grpc.stub.Calls.blockingUnaryCall(Calls.java:129)
	at io.grpc.android.integrationtest.TestServiceGrpc$TestServiceBlockingStub.emptyCall(TestServiceGrpc.java:331)
	at io.grpc.android.integrationtest.IntegrationTester.emptyUnary(IntegrationTester.java:93)
	at io.grpc.android.integrationtest.IntegrationTester.runTest(IntegrationTester.java:72)
	at io.grpc.android.integrationtest.GrpcTestTask.doInBackground(GrpcTestTask.java:53)
	at io.grpc.android.integrationtest.GrpcTestTask.doInBackground(GrpcTestTask.java:12)
	at android.os.AsyncTask$2.call(AsyncTask.java:288)
	at java.util.concurrent.FutureTask.run(FutureTask.java:237)
	at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:231)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
	at java.lang.Thread.run(Thread.java:841)
Caused by: io.grpc.StatusRuntimeException: INTERNAL: Failed starting transport
	at io.grpc.Status.asRuntimeException(Status.java:428)
	at io.grpc.stub.Calls$UnaryStreamToFuture.onClose(Calls.java:324)
	at io.grpc.ChannelImpl$CallImpl$ClientStreamListenerImpl$3.run(ChannelImpl.java:402)
	at io.grpc.SerializingExecutor$TaskRunner.run(SerializingExecutor.java:152)
	... 3 more
Caused by: java.lang.RuntimeException: protocol negotiation failed
	at com.squareup.okhttp.OkHttpTlsUpgrader.upgrade(OkHttpTlsUpgrader.java:95)
	at io.grpc.transport.okhttp.OkHttpClientTransport.start(OkHttpClientTransport.java:293)
	at io.grpc.ChannelImpl.obtainActiveTransport(ChannelImpl.java:199)
	at io.grpc.ChannelImpl.access$600(ChannelImpl.java:55)
	at io.grpc.ChannelImpl$CallImpl.start(ChannelImpl.java:266)
	at io.grpc.stub.Calls.asyncServerStreamingCall(Calls.java:174)
	at io.grpc.stub.Calls.unaryFutureCall(Calls.java:86)
	at io.grpc.stub.Calls.blockingUnaryCall(Calls.java:129)
	at io.grpc.android.integrationtest.TestServiceGrpc$TestServiceBlockingStub.emptyCall(TestServiceGrpc.java:331)
	at io.grpc.android.integrationtest.IntegrationTester.emptyUnary(IntegrationTester.java:93)
	at io.grpc.android.integrationtest.IntegrationTester.runTest(IntegrationTester.java:72)
	at io.grpc.android.integrationtest.GrpcTestTask.doInBackground(GrpcTestTask.java:53)
	at io.grpc.android.integrationtest.GrpcTestTask.doInBackground(GrpcTestTask.java:12)
	at android.os.AsyncTask$2.call(AsyncTask.java:288)
	at java.util.concurrent.FutureTask.run(FutureTask.java:237)
	at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:231)
	... 3 more
]>
	at io.grpc.android.integrationtest.TesterActivityTest.testGrpc(TesterActivityTest.java:65)
	at java.lang.reflect.Method.invokeNative(Native Method)
	at android.test.InstrumentationTestCase.runMethod(InstrumentationTestCase.java:214)
	at android.test.InstrumentationTestCase.runTest(InstrumentationTestCase.java:199)
	at android.test.ActivityInstrumentationTestCase2.runTest(ActivityInstrumentationTestCase2.java:192)
	at android.test.AndroidTestRunner.runTest(AndroidTestRunner.java:191)
	at android.test.AndroidTestRunner.runTest(AndroidTestRunner.java:176)
	at android.test.InstrumentationTestRunner.onStart(InstrumentationTestRunner.java:554)
	at android.app.Instrumentation$InstrumentationThread.run(Instrumentation.java:1701)
```

--- a/okhttp/src/main/java/com/squareup/okhttp/OkHttpTlsUpgrader.java
+++ b/okhttp/src/main/java/com/squareup/okhttp/OkHttpTlsUpgrader.java
@@ -34,6 +34,7 @@ package com.squareup.okhttp;
 import com.google.common.base.Preconditions;
 
 import com.squareup.okhttp.internal.Platform;
+import com.squareup.okhttp.internal.SelectedProtocolQuerier;
 
 import java.io.IOException;
 import java.net.InetSocketAddress;
@@ -75,22 +76,15 @@ public final class OkHttpTlsUpgrader {
 
     Platform platform = Platform.get();
 
-    String negotiatedProtocol = null;
-    try {
-      // It's possible that the user provided SSLSocketFactory has already done the handshake
-      // when creates the SSLSocket.
-      negotiatedProtocol = platform.getSelectedProtocol(sslSocket);
-    } catch (Exception e) {
-      // In some implementations, querying selected protocol before the handshake will fail with
-      // exception.
-    }
+    // It's possible that the user provided SSLSocketFactory has already done the handshake
+    // when creates the SSLSocket.
+    String negotiatedProtocol = SelectedProtocolQuerier.getSelectedProtocol(sslSocket);
     if (negotiatedProtocol == null) {
-
       try {
         // Force handshake.
         sslSocket.startHandshake();
 
-        negotiatedProtocol = platform.getSelectedProtocol(sslSocket);
+        negotiatedProtocol = SelectedProtocolQuerier.getSelectedProtocol(sslSocket);
         if (negotiatedProtocol == null) {
           throw new RuntimeException("protocol negotiation failed");
         }

--- a/okhttp/src/main/java/com/squareup/okhttp/internal/SelectedProtocolQuerier.java
+++ b/okhttp/src/main/java/com/squareup/okhttp/internal/SelectedProtocolQuerier.java
@@ -1,0 +1,82 @@
+/*
+ * Copyright 2014, Google Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *    * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *    * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.squareup.okhttp.internal;
+
+import java.net.Socket;
+
+import javax.net.ssl.SSLSocket;
+
+/**
+ * A helper class located in package com.squareup.okhttp.internal for getting protocol
+ * selected by NPN on Android.
+ */
+public class SelectedProtocolQuerier {
+  private static Platform platform = Platform.get();
+
+  // byte[] getNpnSelectedProtocol()
+  private static final OptionalMethod<Socket> GET_NPN_SELECTED_PROTOCOL =
+      new OptionalMethod<Socket>(byte[].class, "getNpnSelectedProtocol");
+  private static boolean android;
+
+  /** Returns the negotiated protocol, or null if no protocol was negotiated. */
+  public static String getSelectedProtocol(SSLSocket socket) {
+    String protocol = null;
+    try {
+      protocol = platform.getSelectedProtocol(socket);
+    } catch (Exception e) {
+      // In some implementations, querying selected protocol before the handshake will fail with
+      // exception.
+    }
+    if (protocol == null && android && GET_NPN_SELECTED_PROTOCOL.isSupported(socket)) {
+      byte[] result = (byte[]) GET_NPN_SELECTED_PROTOCOL.invokeWithoutCheckedException(socket);
+      if (result != null) {
+        protocol = new String(result, Util.UTF_8);
+      }
+    }
+    return protocol;
+  }
+
+  static {
+    android = true;
+    try {
+      // Attempt to find Android 2.3+ APIs.
+      Class.forName("com.android.org.conscrypt.OpenSSLSocketImpl");
+    } catch (ClassNotFoundException ignored) {
+      try {
+        // Older platform before being unbundled.
+        Class.forName("org.apache.harmony.xnet.provider.jsse.OpenSSLSocketImpl");
+      } catch (ClassNotFoundException ignored2) {
+        android = false;
+      }
+    }
+  }
+}

